### PR TITLE
feat(caimake): add implementation according to specification

### PR DESCRIPTION
**Pull request by zoumo**

### CODE REVIEW

File: docs/Caimake.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+### Installation
+
+#### 1. Download from Github release
+
+Download directly from GIthub [release](directly)
+
+#### 2. Use `go get`
+
+```shell
+go get -u https://github.com/caicloud/build-infra/...
```
Comment (deyuanix) [inline docs/Caimake.md]: no `https`

```
go get -u github.com/caicloud/build-infra/...
```

File: docs/Caimake.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+Getting started from a git repo, use `caimake init` to initialize it. 
+
+```shell
+caimake init
+```
+
+`caimake init` will scan your project:
+
+-   find all subdirectories under `cmd` as targets of `make build` 
+-   find all subdirectories under `build` as targets of `make docker`
```
Comment (deyuanix) [inline docs/Caimake.md]: This is golang convention, not makefile convention.  Can we apply this to other languages like node, python?

File: cmd/caimake/main.go [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/caicloud/build-infra/pkg/caimake/cmd"
+)
+
+func main() {
+	cmd := cmd.NewCaimakeCommand()
```
Comment (deyuanix) [inline cmd/caimake/main.go]: Just curious, are you explicitly overwriting `cmd` here?

File: pkg/caimake/cmd/cmd.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	"github.com/caicloud/build-infra/pkg/basher"
+	"github.com/caicloud/build-infra/pkg/version"
+	"github.com/caicloud/nirvana/cli"
+	"github.com/caicloud/nirvana/log"
+	"github.com/spf13/cobra"
+)
+
+const (
+	dotfile = "~/.caimake"
+	project = "make-rules"
```
Comment (deyuanix) [inline pkg/caimake/cmd/cmd.go]: what does this `project` mean?

File: .caimake [hunk 5]
```diff
@@ -0,0 +1,107 @@
+#!/bin/bash
```
Comment (deyuanix) [inline .caimake]: Add License to all files :)

File: docs/Caimake.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Caimake
```
Comment (deyuanix) [inline docs/Caimake.md:1]: note, we need a lot more explanations for this to be useable.  we can fix this in following PRs.

File: docs/Caimake.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+Download directly from GIthub [release](directly)
+
+#### 2. Use `go get`
+
+```shell
+go get -u https://github.com/caicloud/build-infra/...
+```
+
+### Requirements
```
Comment (deyuanix) [inline docs/Caimake.md]: any version requirement?

File: docs/Caimake.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+-   `make`
+-   `sed`
```
Comment (deyuanix) [inline docs/Caimake.md]: `sed` syntax is different in mac and linux, make sure this works in both environment.

File: docs/Caimake.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+### Usage
+
+Getting started from a git repo, use `caimake init` to initialize it. 
+
+```shell
+caimake init
+```
+
+`caimake init` will scan your project:
```
Comment (deyuanix) [inline docs/Caimake.md]: I think this command will put libraries under ~/.cmake and create a Makefile for your project?

How about running two commands for this purpose, one `caimake init` to init global workspace; then something like `caimake init golang` to init current project.  Thus you only need to run `caimake init` once for all.

File: hack/make-rules/Makefile.go.tmpl [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+# Args:
+#   WHAT: Directory names to build.  If any of these directories has a 'main'
+#     package, the build will produce executable files under bin/.
+#     If not specified, "everything" will be built.
+#   GOFLAGS: Extra flags to pass to 'go' when building.
+#   GOLDFLAGS: Extra linking flags passed to 'go' when building.
+#   GOGCFLAGS: Additional go compile flags passed to 'go' when building.
+#
+# Example:
+#   make
```
Comment (deyuanix) [inline hack/make-rules/Makefile.go.tmpl]: as discussed offline, this doesn't work

File: docs/Caimake.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+```shell
+go get -u https://github.com/caicloud/build-infra/...
+```
+
+### Requirements
+
+-   `make`
+-   `sed`
+-   `git`
+-   `docker` for build and push container
```
Comment (deyuanix) [inline docs/Caimake.md]: for building and pushing container

File: docs/Caimake.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+-   `golang` for compile golang project
```
Comment (deyuanix) [inline docs/Caimake.md]: for compiling golang project

File: docs/Caimake.md (hunk 9 above)
Comment (deyuanix) [reply docs/Caimake.md]: Note you seem to have the command `caimake golang`, is this the one to init a golang project?

File: hack/make-rules/install.sh [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+[[ -d "${PRJ_ROOT}/cmd" ]] && cmds=($(ls ${PRJ_ROOT}/cmd))
+[[ -d "${PRJ_ROOT}/build" ]] && builds=($(ls ${PRJ_ROOT}/build))
+prj=$(basename ${PRJ_ROOT})
+
+"${SED_I[@]}" "s|__CMD_TARGETS__|\$(addprefix cmd/,${cmds[*]-})|g" ${MAKEFILE}
+"${SED_I[@]}" "s|__DOCKER_TARGETS__|\$(addprefix build/,${builds[*]-})|g" ${MAKEFILE}
+"${SED_I[@]}" "s|__PROJECT_PREFIX__|${prj}-|g" ${MAKEFILE}
+
+log::status "Genenrate Makefile successfully on" ${MAKEFILE}
```
Comment (deyuanix) [inline hack/make-rules/install.sh]: type: Generate

File: hack/make-rules/install.sh [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+cd ${MAKR_RULES_ROOT}
+MAKEFILE="${PRJ_ROOT}/Makefile"
+
+if [[ -e ${MAKEFILE} ]]; then
+	if ! log::confirm "${MAKEFILE} already exists, continue to override it?"; then
+		exit 0
+	fi
+fi
+
+cp "${MAKR_RULES_ROOT}/Makefile.${PROJECT_TYPE}.tmpl" ${MAKEFILE}
```
Comment (deyuanix) [inline hack/make-rules/install.sh]: Verify that `PROJECT_TYPE` can only be `golang`?

File: docs/Caimake.md (hunk 8 above)
Comment (deyuanix) [reply docs/Caimake.md]: seems you've handled this.

File: pkg/caimake/cmd/init.go [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			os.Setenv(EnvProjectType, cli.GetString("type"))
+			bash.Run("install.sh", args)
+		},
+	})
+	cmd.AddFlag(
+		cli.StringFlag{
+			Name:     "type",
+			EnvKey:   EnvProjectType,
+			DefValue: "go",
```
Comment (deyuanix) [inline pkg/caimake/cmd/init.go]: make sure you are consistent with `go` and `golang`.  i think in someplace, you are using `go`, and someplace uses `golang`.

File: hack/make-rules/Makefile.go.tmpl [hunk 16]
```diff
@@ -0,0 +1,285 @@
+# Copyright 2017 The Caicloud Authors.
+#
+# Old-skool build tools.
+#
+# Commonly used targets (see each target for more information):
+#   all: Build code.
+#   test: Run tests.
+#   clean: Clean up.
+# 
+# see docs/Makefile.md for more information.
```
Comment (deyuanix) [inline hack/make-rules/Makefile.go.tmpl]: Where is `docs/Makefile.md`?

File: pkg/caimake/cmd/golang.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+import (
+	"github.com/caicloud/nirvana/cli"
+	"github.com/spf13/cobra"
+)
+
+// NewCmdGolang ...
+func NewCmdGolang() *cli.Command {
+	cmd := cli.NewCommand(&cobra.Command{
+		Use:   "go",
+		Short: "make for golang",
```
Comment (deyuanix) [inline pkg/caimake/cmd/golang.go]: Add long description?

**Review by deyuanix — CHANGES_REQUESTED** [review]
First round. nice tool in general.

**Comment by deyuanix** [comment]
@zoumo

**Comment by deyuanix** [comment]
```
$ caimake init
/home/deyuan/.caimake/make-rules/lib/golang.sh: line 39: GO_BUILD_TARGETS[@]: unbound variable
```

Run into :point_up_2: error when executing the command on Linux.

**Comment by deyuanix** [comment]
A general concern about using ~/.caimake as a library is that a project using this build-infra is no longer self-contained. For example, if I switch to another computer and clone the project using build-infra, then `make build` will fail since I don't have ~/.caimake.

It's tempted to say that one can always run `caimake` to init the library again, but then it comes the problem of compatibility.  How do I know which version the original project use?

File: Makefile [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+#   make container
+#   make container WAHT=build/server
+endef
+.PHONY: container 
+ifeq ($(HELP),y)
+container:
+	@echo "$$CONTAINER_HELP_INFO"
+else
+container: check-caimake
+	caimake docker build $(WHAT)
```
Comment (deyuanix) [inline Makefile]: So you call `caimake` here and in `caimake` binary, call bash script?  Why the turnaround?

File: pkg/caimake/cmd/clean.go [hunk 19]
```diff
@@ -0,0 +1,17 @@
+package cmd
+
+import (
+	"github.com/caicloud/nirvana/cli"
+	"github.com/spf13/cobra"
```
Comment (deyuanix) [inline pkg/caimake/cmd/clean.go:5]: nirvana should capsulate all underline information, importing `cobra` here worries me.

File: Makefile (hunk 18 above)
Comment (zoumo) [reply Makefile]: a binary is easier for distribution than a mess scripts.

File: docs/Caimake.md (hunk 8 above)
Comment (zoumo) [reply docs/Caimake.md]: yep

File: docs/Caimake.md (hunk 9 above)
Comment (zoumo) [reply docs/Caimake.md]: SGTM

File: docs/Caimake.md (hunk 2 above)
Comment (zoumo) [reply docs/Caimake.md]: I will move this convention to golang section, and add a new `caimake golang init` command to help init a golang project.

File: hack/make-rules/Makefile.go.tmpl (hunk 16 above)
Comment (zoumo) [reply hack/make-rules/Makefile.go.tmpl]: Embarrassing mistakes

File: hack/make-rules/Makefile.go.tmpl (hunk 10 above)
Comment (zoumo) [reply hack/make-rules/Makefile.go.tmpl]: fixed

**Comment by zoumo** [comment]
caimake will check all assets in `~/.caimake` everytime, make sure that the scripts exist and their sha256sum . You can find the action in `PersistentPreRun` of `camake` cmd
> A general concern about using ~/.caimake as a library is that a project using this build-infra is no longer self-contained. For example, if I switch to another computer and clone the project using build-infra, then make build will fail since I don't have ~/.caimake.
It's tempted to say that one can always run caimake to init the library again, but then it comes the problem of compatibility. How do I know which version doesn't the original project use?

**Comment by zjj2wry** [comment]
```
deploy:
	kubectl set image deploy tenant-admin=$(REGISTRIES)/tenant-admin:$(VERSION) -l addon-manager.caicloud.io/name=tenant-admin -n default
	kubectl rollout status -w $(shell kubectl get deploy -l addon-manager.caicloud.io/name=tenant-admin --no-headers -o name -n default) -n default
```
i have make deploy in my makefile, how can integrate kubectl into caimake?

**Comment by deyuanix** [comment]
@zoumo ping me when ready

**Comment by deyuanix** [comment]
> i have make deploy in my makefile, how can integrate kubectl into caimake?

@zjj2wry ideally, this should be done in cicd pipeline :point_left: @supereagle

**Comment by supereagle** [comment]
Plan to support CD in Q2.

> @zjj2wry ideally, this should be done in cicd pipeline 👈 @supereagle

**Comment by zoumo** [comment]
ping @ddysher

**Comment by zoumo** [comment]
ping  @ddysher

File: pkg/caimake/cmd/init.go [hunk 20]
```diff
@@ -0,0 +1,194 @@
+package cmd
```
Comment (deyuanix) [inline pkg/caimake/cmd/init.go:1]: Use the license tool in nirvana to add license (in another PR)

File: .gitignore [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+*.so
+*.dylib
+
+# Test binary, build with `go test -c`
+*.test
+
+# Output of the go coverage tool, specifically when used with LiteIDE
+*.out
+
+# Project-local glide cache, RE: https://github.com/Masterminds/glide/issues/736
```
Comment (deyuanix) [inline .gitignore:13]: we are using `dep` right?

File: .gitignore [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+.glide/
+.vscode/
+bin/
+caimake/caimake.sh
```
Comment (deyuanix) [inline .gitignore:17]: why do you ignore this?  IIUC, it's put under `/hack/caimake/caimake.sh`.

File: docs/Caimake.md [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+### Initialization
+
+Getting started from a git repo, use `caimake init <language>` to initialize it. 
+The following languages are supported now:
+- go
+
+`caimake init` will scan your project:
+-   treat project's name as docker images prefix
+
+for `golang`, it will:
```
Comment (deyuanix) [inline docs/Caimake.md:40]: You are using `go` in source code, let's use `golang` in all places?

```go
	cmd := cli.NewCommand(&cobra.Command{
		Use:   "go",
		Short: "make for golang",
		Run:   runHelp,
	})
```

File: docs/Caimake.md [hunk 24]
```diff
@@ -0,0 +1,75 @@
...
+-   find all subdirectories under `cmd` as targets of `make build` 
+-   find all subdirectories under `build` as targets of `make container` 
+
+Then, caimake generates a `Makefile` and put `.caimake` under the project root.
+
+Now, you are able to use `make` instead of `caimake`.
+
+Use `caimake [command] —help` for more information about teh command.
+
+See [Makefile Specification](Specification.md) for more information about detail.
```
Comment (deyuanix) [inline docs/Caimake.md:50]: feel like we'd better add some common usage here (we can add it as we go)

File: pkg/caimake/cmd/clean.go [hunk 25]
```diff
@@ -0,0 +1,17 @@
+package cmd
+
+import (
+	"github.com/caicloud/nirvana/cli"
```
Comment (deyuanix) [inline pkg/caimake/cmd/clean.go:4]: This package is gone, should we use cobra directly?

**Review by deyuanix — COMMENTED** [review]
last few high-level comments

File: docs/Caimake.md (hunk 23 above)
Comment (zoumo) [reply docs/Caimake.md:40]: I have already changed all visible `golang` to `go`.
`go` is shorter and more concise than `golang`

File: pkg/caimake/cmd/clean.go (hunk 25 above)
Comment (zoumo) [reply pkg/caimake/cmd/clean.go:4]: I lock `nirvana` with version `v0.1.0-alpha.0` where `cli` pkg is still alive.
I‘d like to replace it util `mamba` rebirth.

File: .gitignore (hunk 21 above)
Comment (zoumo) [reply .gitignore:13]: yes, this dotfile is downloaded from github

File: .gitignore (hunk 22 above)
Comment (zoumo) [reply .gitignore:17]: `/hack/caimake/caimake.sh` MUST be generated every time when building.
It is a all-in-one script containing everything in `/hack/make-rules`

see more imformation in `Makefile.expansion`

File: docs/Caimake.md (hunk 23 above)
Comment (deyuanix) [reply docs/Caimake.md:40]: Then don't use `golang` here (we can change along with readme rewrite)

**Review by deyuanix — APPROVED** [review]

**Pull request merged**